pkg/events/cloud/cloud.go

```diff
@@ -19,81 +19,129 @@ import (
 	"context"
 	"encoding/json"
 	"strings"
+	"sync"
 
+	"go.thethings.network/lorawan-stack/v3/pkg/component"
 	"go.thethings.network/lorawan-stack/v3/pkg/events"
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
 	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
 	"gocloud.dev/pubsub"
 )
 
 // WrapPubSub wraps an existing PubSub and publishes all events received from Go Cloud to that PubSub.
 // If the subURL is an empty string, this PubSub will only publish to Go Cloud.
-func WrapPubSub(ctx context.Context, wrapped events.PubSub, pubURL, subURL string) (ps *PubSub, err error) {
+func WrapPubSub(wrapped events.PubSub, c *component.Component, pubURL, subURL string) (ps *PubSub, err error) {
+	ctx := log.NewContextWithField(c.Context(), "namespace", "events/cloud")
+	ctx, cancel := context.WithCancel(ctx)
 	ps = &PubSub{
 		PubSub:      wrapped,
+		c:           c,
 		ctx:         ctx,
+		cancel:      cancel,
 		contentType: "application/protobuf",
+		subURL:      subURL,
 	}
 	ps.topic, err = pubsub.OpenTopic(ctx, pubURL)
 	if err != nil {
 		return nil, err
 	}
-	if subURL != "" {
-		ps.subscription, err = pubsub.OpenSubscription(ctx, subURL)
-		if err != nil {
-			return nil, err
-		}
-		go func() {
-			for ctx.Err() == nil {
-				msg, err := ps.subscription.Receive(ctx)
-				if err != nil {
-					return
-				}
-				switch msg.Metadata["content-type"] {
-				case "application/protobuf":
-					var evtpb ttnpb.Event
-					if err = evtpb.Unmarshal(msg.Body); err == nil {
-						if evt, err := events.FromProto(&evtpb); err == nil {
-							ps.PubSub.Publish(evt)
-						}
-					}
-				case "application/json":
-					if evt, err := events.UnmarshalJSON(msg.Body); err == nil {
-						ps.PubSub.Publish(evt)
-					}
-				}
-				msg.Ack()
-			}
-		}()
-	}
 	return ps, nil
 }
 
 // NewPubSub creates a new PubSub that publishes and subscribes to Go Cloud.
 // If the subURL is an empty string, this PubSub will only publish to Go Cloud.
-func NewPubSub(ctx context.Context, pubURL, subURL string) (*PubSub, error) {
-	return WrapPubSub(ctx, events.NewPubSub(events.DefaultBufferSize), pubURL, subURL)
+func NewPubSub(c *component.Component, pubURL, subURL string) (*PubSub, error) {
+	return WrapPubSub(events.NewPubSub(events.DefaultBufferSize), c, pubURL, subURL)
 }
 
 // PubSub with Go Cloud backend.
 type PubSub struct {
 	events.PubSub
 
-	ctx          context.Context
-	contentType  string
-	topic        *pubsub.Topic
-	subscription *pubsub.Subscription
+	c           *component.Component
```

Comment:
While I'm not a fan of the naming scheme above, I consider those nits.
This one is too important to be considered a nit. `PubSub.c` gives exactly 0 context on what this field is for. Might as well call it `x`, `y`, `foo` or `field`.
```suggestion
	component   *component.Component
```